frappe/oauth.py

```diff
@@ -373,32 +479,77 @@ def validate_user_match(self, id_token_hint, scopes, claims, request):
 		:rtype: True or False
 
 		Method is used by:
-		    - OpenIDConnectAuthCode
-		    - OpenIDConnectImplicit
-		    - OpenIDConnectHybrid
+		- OpenIDConnectAuthCode
+		- OpenIDConnectImplicit
+		- OpenIDConnectHybrid
 		"""
-		if id_token_hint and id_token_hint == frappe.db.get_value("User Social Login", {"parent":frappe.session.user, "provider": "frappe"}, "userid"):
+		if id_token_hint:
+			try:
+				user = None
+				payload = jwt.decode(
+					id_token_hint,
+					options={
+						"verify_signature": False,
+						"verify_aud": False,
+					},
+				)
+				client_id, client_secret = frappe.get_value(
+					"OAuth Client",
+					payload.get("aud"),
+					["client_id", "client_secret"],
+				)
+
+				if payload.get("sub") and client_id and client_secret:
+					user = frappe.db.get_value(
+						"User Social Login",
+						{"userid": payload.get("sub"), "provider": "frappe"},
+						"parent",
+					)
+					user = frappe.get_doc("User", user)
+					verified_payload = jwt.decode(
+						id_token_hint,
+						key=client_secret,
+						audience=client_id,
+						algorithm="HS256",
+						options={
+							"verify_exp": False,
+						},
+					)
+
+					if verified_payload:
+						return user.name == frappe.session.user
+
+			except Exception:
+				return False
```

Comment:
Should we generate error logs on any failures? I can't remember if Frappe automatically generates one on handled exceptions.

Reply:
`validate_user_match` expects to return True or False. https://oauthlib.readthedocs.io/en/latest/oauth2/oidc/validator.html#oauthlib.openid.RequestValidator.validate_user_match

in normal cases if `frappe.session.user != "Guest"` and we're good to go.

in case `id_token_hint` is passed to authorization request it should verify user from that valid/invalid/illegal token.
In that case whatever fails should return False, No need to log these errors. It'll be just unnecessary logs if illegal id_tokens are used.

Reply:
I see, fair enough